crates/driver/src/driver.rs

```diff
@@ -34,14 +42,62 @@ impl Driver {
         &self,
         auction: AuctionWithId,
     ) -> Result<SettlementSummary, SolveError> {
-        let fetch_liquidity_from_block = block_number(&self.block_stream.borrow())?;
-        let auction = self
-            .auction_converter
-            .convert_auction(auction, fetch_liquidity_from_block)
-            .await?;
-        let summary = self.solver.commit(auction).await?;
-        tracing::info!(?summary, "computed winning settlement summary");
-        Ok(summary)
+        Self::solve_until_deadline(
+            auction,
+            self.solver.clone(),
+            self.auction_converter.clone(),
+            self.block_stream.clone(),
+            // TODO get deadline from autopilot auction
+            Instant::now() + Duration::from_secs(25),
+        )
+        .await
+        .map_err(SolveError::from)
+    }
+
+    /// Computes a solution with the liquidity collected from a given block.
+    async fn compute_solution_for_block(
+        auction: AuctionWithId,
+        block: Block,
+        converter: Arc<dyn AuctionConverting>,
+        solver: Arc<dyn CommitRevealSolving>,
+    ) -> Result<SettlementSummary> {
+        let block = block_number(&block)?;
+        let auction = converter.convert_auction(auction, block).await?;
+        solver.commit(auction).await
+    }
+
+    /// Keeps solving the given auction with the most recent liquidity at that time or until the
+    /// auction deadline is reached.
+    /// This function uses a `WatchStream` to get notified about new blocks which will start with
+    /// yielding the current block immediately.
+    async fn solve_until_deadline(
+        auction: AuctionWithId,
+        solver: Arc<dyn CommitRevealSolving>,
+        converter: Arc<dyn AuctionConverting>,
+        block_stream: CurrentBlockStream,
+        deadline: Instant,
+    ) -> Result<SettlementSummary> {
+        last_completed(
+            |block| {
+                Box::pin(
+                    Self::compute_solution_for_block(
+                        auction.clone(),
+                        block,
+                        converter.clone(),
+                        solver.clone(),
+                    )
+                    .fuse(),
+                )
+            },
+            into_stream(block_stream).fuse(),
+            tokio::time::sleep_until(deadline.into()).fuse(),
+        )
+        .await
+        .unwrap_or_else(|| {
+            Err(anyhow::anyhow!(
+                "could not compute a result before the deadline"
+            ))
+        })
```

Comment:
```suggestion
        .context("could not compute a result before the deadline")
```

Reply:
Not quite. Since `last_completed()` returns an `Option<Result<SettlementSummary>>` just using `context()` would produce `Result<Result<SettlementSummary>>`.
Throughout this whole PR type inference was a huge problem.

Reply:
Ah! Right - you should be able to `flatten` the result then (if the error types are the same, which I think they are).

Reply:
Nope, it's an unstable feature. 😅

Reply:
😓